aten/src/ATen/native/TopKImpl.h

```diff
@@ -49,36 +49,60 @@ void topk_impl_loop(
       if (largest) {
         std::partial_sort(queue.begin(), queue.begin() + k, queue.end(),
           [](const elem_t& x, const elem_t& y) -> bool {
-            return ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first));
+            if ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first)) return true;
+            else if (!(x.first >= y.first)) return false;
+            return x.second < y.second;
           });
       } else {
         std::partial_sort(queue.begin(), queue.begin() + k, queue.end(),
           [](const elem_t& x, const elem_t& y) -> bool {
-            return ((!_isnan<accscalar_t>(x.first) && _isnan<accscalar_t>(y.first)) || (x.first < y.first));
+            if ((!_isnan<accscalar_t>(x.first) && _isnan<accscalar_t>(y.first)) || (x.first < y.first)) return true;
+            else if (!(x.first <= y.first)) return false;
+            return x.second < y.second;
           });
       }
     } else {
       if (largest) {
-        std::nth_element(queue.begin(), queue.begin() + k - 1, queue.end(),
+        if (stable) {
+          std::sort(queue.begin(), queue.end(),
+          [](const elem_t& x, const elem_t& y) -> bool {
+            if ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first)) return true;
+            else if (!(x.first >= y.first)) return false;
+            return x.second < y.second;
+          });
+        }
+        else {
+          std::nth_element(queue.begin(), queue.begin() + k - 1, queue.end(),
           [](const elem_t& x, const elem_t& y) -> bool {
             return ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first));
           });
-        if (sorted) {
-          std::sort(queue.begin(), queue.begin() + k - 1,
-            [](const elem_t& x, const elem_t& y) -> bool {
-              return ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first));
-            });
+          if (sorted) {
+            std::sort(queue.begin(), queue.begin() + k - 1,
+              [](const elem_t& x, const elem_t& y) -> bool {
+                return ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first));
+              });
+          }
         }
       } else {
-        std::nth_element(queue.begin(), queue.begin() + k -1, queue.end(),
+        if (stable) {
+          std::sort(queue.begin(), queue.end(),
+          [](const elem_t& x, const elem_t& y) -> bool {
+            if ((!_isnan<accscalar_t>(x.first) && _isnan<accscalar_t>(y.first)) || (x.first < y.first)) return true;
+            else if (!(x.first <= y.first)) return false;
+            return x.second < y.second;
+          });
```

Comment:
Hmm, why can't we directly dispatch to `at::sort` which already handles `stable=True`?